contracts/GatewayFetcher.sol

```diff
@@ -53,64 +53,79 @@ library GatewayFetcher {
+	function addSlot(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_SLOT_ADD); }
+	//function addSlot(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.pushSlot().add().slot(); } 
+	function slot(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_SLOT); }
+	function setSlot(GatewayRequest memory r, uint256 x) internal pure returns (GatewayRequest memory) { return r.push(x).slot(); }
+	function offset(GatewayRequest memory r, uint256 dx) internal pure returns (GatewayRequest memory) { return r.push(dx).addSlot(); }
+	function pushSlot(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_PUSH_SLOT); }
+	//function zeroSlot(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_SLOT_ZERO); } // 20240922: deprecated
+	
 	function setOutput(GatewayRequest memory r, uint8 i) internal pure returns (GatewayRequest memory) { return r.addByte(OP_SET_OUTPUT).addByte(i); }
 	
-	function read(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_READ_SLOTS).addByte(1); }
 	function read(GatewayRequest memory r, uint8 n) internal pure returns (GatewayRequest memory) { return r.addByte(OP_READ_SLOTS).addByte(n); }
 	function readBytes(GatewayRequest memory r) internal pure returns (GatewayRequest memory) { return r.addByte(OP_READ_BYTES); }
 	function readArray(GatewayRequest memory r, uint8 step) internal pure returns (GatewayRequest memory) { return r.addByte(OP_READ_ARRAY).addByte(step); }
 	
-	function push(GatewayRequest memory r, uint256 x) internal pure returns (GatewayRequest memory) { return push(r, abi.encode(x)); }
-	function push(GatewayRequest memory r, address x) internal pure returns (GatewayRequest memory) { return push(r, abi.encode(x)); }
-	function push(GatewayRequest memory r, bytes32 x) internal pure returns (GatewayRequest memory) { return push(r, abi.encode(x)); }
+	function push(GatewayRequest memory r, bytes32 x) internal pure returns (GatewayRequest memory) { return r.push(uint256(x)); }
+	function push(GatewayRequest memory r, address x) internal pure returns (GatewayRequest memory) { return r.push(abi.encode(x)); }
 	function push(GatewayRequest memory r, string memory s) internal pure returns (GatewayRequest memory) { return push(r, bytes(s)); }
-	function push(GatewayRequest memory r, GatewayRequest memory x) internal pure returns (GatewayRequest memory) { return push(r, x.encode()); }
+	function push(GatewayRequest memory r, GatewayRequest memory p) internal pure returns (GatewayRequest memory) { return push(r, p.encode()); }
+
+	function push(GatewayRequest memory r, uint256 x) internal pure returns (GatewayRequest memory) {
+		return x < 256 ? r.addByte(OP_PUSH_BYTE).addByte(uint8(x)) : r.push(abi.encode(x)); 
```

Comment:
yes if I understand correctly, `OP_PUSH_BYTE` expands back to 32 bytes when read from the program

Reply:
the width of integers is important because `follow()` (mapping lookups) use bytes of variable length

`keccak(concat(uint8(1), slot))` is different from `keccak(concat(uint256(1), slot))`

the new unary math ops allow you to cast using `.push(0).or()` or `push(0).plus()`
* if you have `0xAB`, this results in `0x000000000000000000000000000000AB`
* if you have something longer than 32 bytes, it truncates, as if you did `bytes32(...)`